File: clients/typescript/src/simple.ts
import type { AddOptions, AddResult, SearchOptions, EntitiesOptions, RelatedOptions, CompressOptions, CompressResult, MemDogConfig } from "./types.js";
import { MemDogClient } from "./client.js";

/**
 * High-level 7-method facade for the memdog API.
 *
 * For full API coverage (~80 methods) use {@link MemDogClient} directly,
 * accessible via the {@link MemDog.client} property.
 *
 * @example
 * ```ts
 * const m = new MemDog({ baseUrl: "http://localhost:8080", apiKey: "key" });
 * const { dataId } = await m.add("Hello world", { tags: ["greeting"] });
 * const results = await m.search("hello", { limit: 5 });
 * await m.delete(dataId);
 *
 * // Access the full client for advanced operations
 * await m.client.semanticSearch("hello", { searchMode: "hybrid" });
 * ```
 */
export class MemDog {
  private readonly _client: MemDogClient;
  private readonly userId?: string;

  constructor(config: MemDogConfig) {
    this._client = new MemDogClient(config);
    this.userId = config.userId;
  }

  /** Access the full MemDogClient for advanced operations. */
  get client(): MemDogClient {
    return this._client;
  }

  // ---------------------------------------------------------------------------
  // add
  // ---------------------------------------------------------------------------

  async add(content?: string, opts: AddOptions = {}): Promise<AddResult> {
    const uid = opts.userId ?? this.userId;
    const createOpts: Parameters<MemDogClient["createData"]>[0] = {};
    if (content !== undefined) createOpts.content = content;
    if (opts.file) createOpts.file = opts.file;
    if (opts.name) createOpts.name = opts.name;
    if (opts.description) createOpts.description = opts.description;
    if (opts.tags?.length) createOpts.tags = opts.tags;

    const memIds: string[] = [];
    if (opts.memoryId) memIds.push(opts.memoryId);
    if (opts.memoryIds?.length) memIds.push(...opts.memoryIds);
    if (memIds.length) createOpts.memoryIds = memIds;

    const data = await this._client.createData(createOpts);
    const dataId = (data.data_id ?? data.id) as string;
    const result: AddResult = { dataId, memoryId: opts.memoryId };

    if (opts.memoryType && !opts.memoryId) {
      const today = new Date().toISOString().slice(0, 10);
      const memName = `auto-${opts.memoryType}-${today}`;
      let mid = await this.findAutoMemory(opts.memoryType, memName, uid);

      if (!mid) {
        const memPayload: Record<string, unknown> = { memory_type: opts.memoryType, name: memName };
        if (uid) memPayload.user_id = uid;
        const memData = await this._client.createMemory({
          memoryType: opts.memoryType,
          name: memName,
          userId: uid,
        });
        mid = (memData.memory_id ?? memData.id) as string;
      }

      if (mid && dataId) {
        await this._client.addDataToMemory(mid, [dataId]);
      }
      result.memoryId = mid;
    }

    return result;
  }

  private async findAutoMemory(memoryType: string, name: string, userId?: string): Promise<string | undefined> {
    try {
      const data = await this._client.listMemories({ memoryType, limit: 50, userId });
      const items: Record<string, unknown>[] = Array.isArray(data)
        ? data
        : ((data as Record<string, unknown>).items as Record<string, unknown>[]) ?? [];
      for (const mem of items) {
        if (mem.name === name) return (mem.memory_id ?? mem.id) as string;
      }
    } catch { /* swallow */ }
    return undefined;
  }

  // ---------------------------------------------------------------------------
  // search
  // ---------------------------------------------------------------------------

  async search(query: string, opts: SearchOptions = {}): Promise<Record<string, unknown>[]> {
    const uid = opts.userId ?? this.userId;
    const limit = opts.limit ?? 10;

    if (opts.useAi) {
      const data = await this._client.aiQuery(query, { memoryIds: opts.memoryIds });
      return Array.isArray(data) ? data : [data as Record<string, unknown>];
    }

    if (opts.memoryType) {
      const data = await this._client.listMemories({ memoryType: opts.memoryType, limit, userId: uid });
      const items: Record<string, unknown>[] = Array.isArray(data)
        ? data
        : ((data as Record<string, unknown>).items as Record<string, unknown>[]) ?? [];
      return items.slice(0, limit);
    }

    const data = await this._client.listUserData({ user: uid, format: "meta", limit });
    const items: Record<string, unknown>[] = Array.isArray(data)
      ? data
      : ((data as Record<string, unknown>).items as Record<string, unknown>[]) ?? [];
    return items;
  }

  // ---------------------------------------------------------------------------
  // get
  // ---------------------------------------------------------------------------

  async get(dataId: string, version?: number): Promise<Record<string, unknown>> {
    const content = await this._client.getData(dataId, { version });
    let meta: Record<string, unknown> = {};
    try { meta = (await this._client.getMetadata(dataId)) as Record<string, unknown>; } catch { /* optional */ }
    return { dataId, ...meta, content };
  }

  // ---------------------------------------------------------------------------
  // delete
  // ---------------------------------------------------------------------------

  async delete(dataId: string): Promise<boolean> {
    await this._client.deleteData(dataId);
    return true;
  }

  // ---------------------------------------------------------------------------
  // entities
  // ---------------------------------------------------------------------------

  async entities(query: string, opts: EntitiesOptions = {}): Promise<Record<string, unknown>[]> {
    return this._client.searchEntities(query, {
      userId: opts.userId ?? this.userId,
      entityType: opts.entityType,
      limit: opts.limit,
    });
  }

  // ---------------------------------------------------------------------------
  // related
  // ---------------------------------------------------------------------------

  async related(dataId: string, opts: RelatedOptions = {}): Promise<Record<string, unknown>[]> {
    return this._client.getDataEntities(dataId, { userId: opts.userId ?? this.userId });
  }

  // ---------------------------------------------------------------------------
  // compress
  // ---------------------------------------------------------------------------

  async compress(memoryId: string, opts: CompressOptions = {}): Promise<CompressResult> {
    return (await this._client.compressMemory(memoryId, {
      archiveOriginals: opts.archiveOriginals,
      maxSummaryLength: opts.maxSummaryLength,
      userId: opts.userId ?? this.userId,
    })) as CompressResult;
  }
}

File: clients/rust/src/simple.rs
use serde_json::Value;
use crate::client::MemDogClient;
use crate::types::*;

/// High-level 7-method facade. For full API coverage use [`MemDogClient`].
pub struct MemDog {
    inner: MemDogClient,
    user_id: Option<String>,
}

impl MemDog {
    pub fn new(config: MemDogConfig) -> Self {
        Self {
            user_id: config.user_id.clone(),
            inner: MemDogClient::new(config),
        }
    }

    /// Access the full MemDogClient for advanced operations.
    pub fn client(&self) -> &MemDogClient { &self.inner }

    fn resolve_uid<'a>(&'a self, o: &'a Option<String>) -> Option<&'a str> {
        o.as_deref().or(self.user_id.as_deref())
    }

    pub async fn add(&self, content: Option<&str>, opts: AddOptions) -> Result<AddResult> {
        let uid = opts.user_id.as_deref().or(self.user_id.as_deref()).map(|s| s.to_string());
        let create_opts = CreateDataOptions {
            file: opts.file,
            file_name: opts.file_name,
            tags: opts.tags,
            name: opts.name,
            description: opts.description,
            memory_ids: opts.memory_id.as_ref().map(|id| vec![id.clone()]),
            ..Default::default()
        };

        let res = self.inner.create_data(content, create_opts).await?;
        let data_id = res.data_id.clone();
        let mut memory_id = opts.memory_id.clone();

        if opts.memory_type.is_some() && opts.memory_id.is_none() {
            let mtype = opts.memory_type.as_deref().unwrap();
            let today = chrono::Local::now().format("%Y-%m-%d").to_string();
            let mem_name = format!("auto-{}-{}", mtype, today);

            let existing = self.find_auto_memory(mtype, &mem_name, uid.as_deref()).await;
            let mid = if let Some(existing_mid) = existing {
                existing_mid
            } else {
                let mem_data = self.inner.create_memory(CreateMemoryOptions {
                    memory_type: mtype.to_string(),
                    name: mem_name,
                    user_id: uid.clone(),
                    ..Default::default()
                }).await?;
                mem_data.get("memory_id").or_else(|| mem_data.get("id"))
                    .and_then(|v| v.as_str()).unwrap_or("").to_string()
            };

            if !mid.is_empty() && !data_id.is_empty() {
                let _ = self.inner.add_data_to_memory(&mid, &[data_id.clone()]).await;
            }
            memory_id = Some(mid);
        }

        Ok(AddResult { data_id, memory_id })
    }

    async fn find_auto_memory(&self, memory_type: &str, name: &str, user_id: Option<&str>) -> Option<String> {
        let data = self.inner.list_memories(ListMemoriesOptions {
            memory_type: Some(memory_type.to_string()),
            limit: Some(50),
            user_id: user_id.map(|s| s.to_string()),
            ..Default::default()
        }).await.ok()?;

        let items = if data.is_array() {
            data.as_array().cloned().unwrap_or_default()
        } else {
            data.get("items").and_then(|v| v.as_array()).cloned().unwrap_or_default()
        };

        for mem in items {
            if mem.get("name").and_then(|v| v.as_str()) == Some(name) {
                return mem.get("memory_id").or_else(|| mem.get("id"))
                    .and_then(|v| v.as_str()).map(|s| s.to_string());
            }
        }
        None
    }

    pub async fn search(&self, query: &str, opts: SearchOptions) -> Result<Vec<Value>> {
        let uid = self.resolve_uid(&opts.user_id).map(|s| s.to_string());
        let limit = opts.limit.unwrap_or(10);

        if opts.use_ai {
            let data = self.inner.ai_query(query, AIQueryOptions { memory_ids: opts.memory_ids, ..Default::default() }).await?;
            return Ok(if data.is_array() { data.as_array().cloned().unwrap_or_default() } else { vec![data] });
        }

        if let Some(ref mtype) = opts.memory_type {
            let data = self.inner.list_memories(ListMemoriesOptions {
                memory_type: Some(mtype.clone()), limit: Some(limit), user_id: uid, ..Default::default()
            }).await?;
            let items = if data.is_array() { data.as_array().cloned().unwrap_or_default() }
                        else { data.get("items").and_then(|v| v.as_array()).cloned().unwrap_or_default() };
            return Ok(items.into_iter().take(limit as usize).collect());
        }

        let data = self.inner.list_data(ListDataOptions { limit: Some(limit), user: uid, ..Default::default() }).await?;
        let items = if data.is_array() { data.as_array().cloned().unwrap_or_default() }
                    else { data.get("items").and_then(|v| v.as_array()).cloned().unwrap_or_default() };
        Ok(items)
    }

    pub async fn get(&self, data_id: &str, version: Option<i32>) -> Result<Value> {
        let content = self.inner.get_data(data_id, version).await?;
        let meta = self.inner.get_metadata(data_id).await.unwrap_or(Value::Object(serde_json::Map::new()));
        let mut result = if let Value::Object(map) = meta { map } else { serde_json::Map::new() };
        result.insert("data_id".to_string(), Value::String(data_id.to_string()));
        result.insert("content".to_string(), content);
        Ok(Value::Object(result))
    }

    pub async fn delete(&self, data_id: &str) -> Result<bool> {
        self.inner.delete_data(data_id).await?;
        Ok(true)
    }

    pub async fn entities(&self, query: &str, opts: EntitiesOptions) -> Result<Vec<Value>> {
        let data = self.inner.search_entities(query, SearchEntitiesOptions {
            user_id: opts.user_id.or_else(|| self.user_id.clone()),
            entity_type: opts.entity_type,
            limit: opts.limit,
        }).await?;
        Ok(if data.is_array() { data.as_array().cloned().unwrap_or_default() } else { vec![] })
    }

    pub async fn related(&self, data_id: &str, opts: RelatedOptions) -> Result<Vec<Value>> {
        let uid = self.resolve_uid(&opts.user_id).map(|s| s.to_string());
        let data = self.inner.get_data_entities(data_id, uid.as_deref()).await?;
        Ok(if data.is_array() { data.as_array().cloned().unwrap_or_default() } else { vec![] })
    }

    pub async fn compress(&self, memory_id: &str, opts: CompressOptions) -> Result<Value> {
        self.inner.compress_memory(memory_id, CompressMemoryOptions {
            archive_originals: opts.archive_originals,
            max_summary_length: opts.max_summary_length,
            user_id: opts.user_id.or_else(|| self.user_id.clone()),
        }).await
    }
}
